plugins/by-name/treesitter/default.nix

```diff
@@ -365,46 +233,110 @@ lib.nixvim.plugins.mkNeovimPlugin {
-      # NOTE: Upstream state that the parser MUST be at the beginning of runtimepath.
-      # Otherwise the parsers from Neovim takes precedent, which may be incompatible with some queries.
-      (optionalString (cfg.settings.parser_install_dir != null) ''
-        vim.opt.runtimepath:prepend(${lib.nixvim.toLuaObject cfg.settings.parser_install_dir})
+  extraConfig =
+    cfg: opt:
+    let
+      # TODO: Added 2025-12-15 Check both legacy and new api options
+      # Add warning after transition period
+      # Map parser_install_dir to install_dir for main branch compatibility
+      # The main branch uses install_dir, while master branch uses parser_install_dir
+      mainBranchSettings =
+        if cfg.settings ? install_dir then
+          # User provided install_dir directly, use it
+          { inherit (cfg.settings) install_dir; }
+        else if (cfg.settings.parser_install_dir or null) != null then
+          # Map parser_install_dir to install_dir for main branch
+          { install_dir = cfg.settings.parser_install_dir; }
+        else
+          { };
+
+      # TODO: Added 2025-12-15 Check both legacy and new api options
+      # Add warning after transition period
+      # Top-level options (cfg.highlight.enable, cfg.indent.enable) are for main branch
+      # For master branch compatibility, users should use settings.highlight.enable directly
+      highlightEnabled = cfg.highlight.enable || (cfg.settings.highlight.enable or false);
+      indentEnabled = cfg.indent.enable || (cfg.settings.indent.enable or false);
+    in
+    {
+      # TODO: Added 2025-12-15 Runtime detection to support legacy api
+      # Should be removed after transition period
+      # Runtime API detection and conditional setup
+      plugins.treesitter.luaConfig.content = ''
+        -- Detect which branch of nvim-treesitter is installed
+        local has_configs_module = pcall(require, 'nvim-treesitter.configs')
+
+        if has_configs_module then
+          -- Master branch: Use the traditional configs API
+          require('nvim-treesitter.configs').setup(${lib.nixvim.toLuaObject cfg.settings})
+        else
+          -- Main branch: Pass install_dir to setup() and enable features via autocommands
+      ''
+      + (optionalString (mainBranchSettings != { }) ''
+        require'nvim-treesitter'.setup(${lib.nixvim.toLuaObject mainBranchSettings})
+      '')
+      + ''
+
+        -- Enable features via autocommands for main branch
+      ''
+      + (optionalString highlightEnabled ''
+        vim.api.nvim_create_autocmd('FileType', {
+          pattern = '*',
+          callback = function()
+            pcall(vim.treesitter.start)
+          end,
+        })
+      '')
+      + (optionalString indentEnabled ''
+        vim.api.nvim_create_autocmd('FileType', {
+          pattern = '*',
+          callback = function()
+            vim.bo.indentexpr = "v:lua.require'nvim-treesitter'.indentexpr()"
+          end,
+        })
```

Comment:
These autocmds are targeting the same event/pattern/etc. We could probably combine them?

```nix
optionalString (highlightEnabled || indentEnabled) ''
  -- Enable features via autocommands for main branch
    vim.api.nvim_create_autocmd('FileType', {
    pattern = '*',
    callback = function()
      ${optionalString highlightEnabled ''
        pcall(vim.treesitter.start)
      ''}
      ${optionalString indentEnabled ''
        vim.bo.indentexpr = "v:lua.require'nvim-treesitter'.indentexpr()"
      ''}
    end,
  })
''
```

Reply:
Also: I think best practice is to associate the autocmd with an autocmd group? This allows runtime introspection and disablement, IIUC.

For indent, you use `nixvim_treesitter` below. IDK if that should be shared for these autocmds, or if we should have separate groups like `nixvim_treesitter_indent`?

Reply:
> Also: I think best practice is to associate the autocmd with an autocmd group? This allows runtime introspection and disablement, IIUC.

Did that but our ordering of the generated init.lua for augroups means they aren't available for the module code and I'd have to do a custom augroup in this module. If we care that much about the grouping, we can shift from using our nixvim module to doing it in lua.